src/api-caller/api-caller.spec.js

```diff
@@ -0,0 +1,229 @@
+      expect(headers).to.have.property('Accept').equal('application/vnd.api+json')
+    })
+    it('should have a valid `content-type` header', () => {
+      const headers = api._buildHeaders()
+      expect(headers).to.have.property('Content-Type').equal('application/vnd.api+json')
+    })
+  })
+
+  describe('_marshalQuery', () => {
+    it('should properly marshal primitives passed to query params', () => {
+      const query = {
+        foo: 'bar',
+        abc: 123
+      }
+      const result = api._marshalQuery(query)
+
+      expect(result).to.equal('foo=bar&abc=123')
+    })
+    it('should properly marshal query param passed as an flat array', () => {
+      const query = {
+        foo: ['fizz', 'buzz', 'bar'],
+        abc: [123, 'qq', 'test']
+      }
+      const result = api._marshalQuery(query)
+
+      expect(result).to.equal('foo=fizz,buzz,bar&abc=123,qq,test')
+    })
+    it('should properly marshal query param passed as an flat object', () => {
+      const query = {
+        filter: {
+          foo: 'bar',
+          fizz: 'buz'
+        },
+        page: {
+          limit: 20,
+          number: 3
+        }
+      }
+      const result = api._marshalQuery(query)
+
+      expect(result).to.equal(
+        'filter[foo]=bar&filter[fizz]=buz&page[limit]=20&page[number]=3'
+      )
+    })
+    it('should throw an error when nested array provided', () => {
+      const query = {
+        foo: ['fizz', ['buzz', 'bar']]
+      }
+
+      expectThrow(api._marshalQuery(query))
+    })
+    it('should throw an error when nested object provided', () => {
+      const query = {
+        foo: {
+          bar: {
+            fizz: 'buzz'
+          }
+        }
+      }
+
+      expectThrow(api._marshalQuery(query))
+    })
+  })
+
+  describe('_unmarshalResponse', () => {
+    it('should properly deserialize the response data', () => {
+      const rawResponse = {
+        data: {
+          type: 'articles',
+          id: '1',
+          attributes: {
+            title: 'foo',
+            text: 'bar'
+          },
+          relationships: {
+            author: {
+              data: {
+                id: 1,
+                type: 'authors'
+              }
+            }
+          }
+        },
+        included: [{
+          id: 1,
+          type: 'authors',
+          attributes: {
+            first_name: 'John',
+            last_name: 'Doe',
+            age: 45
+          }
+        }]
+      }
+
+      const result = api._unmarshalResponse(rawResponse)
+
+      expect(result).to.have.property('data').deep.equal({
+        resourceId: '1',
+        resourceType: 'authors',
+        title: 'foo',
+        text: 'bar',
+        author: {
+          firstName: 'John',
+          lastName: 'Doe',
+          age: 45
+        }
+      })
+    })
+  })
+})
+
+// TODO: open questions
+// 1. how to define if we need to sign the request
+// 2. how to save the passphrase
+// 3. how to force using horizon timestamp
```

Comment:
auto check of the preinitialized config + autocall of init if dont + init method